File: Merge Sort/MergeSort.py
def Merge(arr, start, end, mid):
    i = start
    j = mid + 1
    temp = []

    # Merging the two halves
    while i <= mid and j <= end:
        if arr[i] <= arr[j]:
            temp.append(arr[i])
            i += 1
        else:
            temp.append(arr[j])
            j += 1

    # Adding remaining elements from the left half (if any)
    while i <= mid:
        temp.append(arr[i])
        i += 1

    # Adding remaining elements from the right half (if any)
    while j <= end:
        temp.append(arr[j])
        j += 1

    # Copying sorted elements back into the original array
    for p in range(len(temp)):
        arr[start + p] = temp[p]

def MergeSort(arr, start, end):
    if start < end:
        mid = (start + end) // 2
        MergeSort(arr, start, mid)
        MergeSort(arr, mid + 1, end)
        Merge(arr, start, end, mid)

def main():
    size = int(input("Enter the size of the array: "))
    arr = []
    print(f"Enter {size} elements in the array:")
    for _ in range(size):
        arr.append(int(input()))
    
    print(f"Elements before Merge Sort: {arr}")
    MergeSort(arr, 0, size - 1)
    print(f"Elements after Merge Sort: {arr}")

if __name__ == "__main__":
    main()

File: Merge Sort/MergeSort.cpp

#include<iostream>
  using namespace std;
  void Merge(int *arr, int start, int end, int mid){
    int i,j,k;
    i=start, j=mid+1, k=start;
    int temp[end+1];
    while(i<=mid && j<=end){
        if(arr[i] <= arr[j]){
            temp[k] = arr[i];
            i++,k++;
        }
        else{
            temp[k] = arr[j];
            j++,k++;
        }
    }
    while(i<=mid){
        temp[k] = arr[i];
        i++,k++;
    }
    while(j<=end){
        temp[k] = arr[j];
        j++,k++;
    }
    for(int p=start; p<=end;p++){
        arr[p] = temp[p];
    }
  }

  void MergeSort(int *array, int start, int end){
    if(start<end){
        int mid = (start+end)/2;
        MergeSort(array,start,mid);
        MergeSort(array,mid+1,end);
        Merge(array, start, end, mid);
    }
  }
  
int main() {
    int size;
    cout << "Enter size of an array: ";
    cin >> size; 

    // Dynamically allocate an array of the given size
    int *arr = new int[size]; 
    cout << "Enter " << size << " elements in an array:";
    for(int i = 0; i < size; i++) {
        cin >> arr[i]; // Read elements into the array
    }

    cout << "Elements before Merge sorting:" << endl;
    for(int i = 0; i < size; i++) {
        cout << arr[i] << " "; // Print elements before sorting
    }
    cout << endl;

    MergeSort(arr,0,(size-1)); // Call Merge sort function to sort the array

    cout << "Elements after Merge sorting:" << endl;
    for(int i = 0; i < size; i++) {
        cout << arr[i] << " "; // Print elements after sorting
    }
    cout << endl;

    delete [] arr; // Deallocate the dynamically allocated array
    return 0; // Return 0 to indicate successful execution
}